### DISCUSSION TITLE
Plugin / Adapter Architecture

### DISCUSSION THREAD

**Discussion by thescientist13** [ProjectEvergreen/wcc#90] [Ideas]
## Overview

As part of introducing #69, it got me thinking about how WCC could be tapped into to help others deliver or augment the AST generation done by WCC.  WCC could help facilitate getting to the output state, allowing anyone to build a custom transform on top of it.

## Benefits
The overall benefit that come to mind would be:
1. Sharing the DOM Shim implementation
1. The SSR mechanism

### Custom Specification
Custom walker for any implementation you want, like for a SFC (Single File Component)
```html
<style>
:host {
  color: red;
  background-color: blue;
}
</style>

<script>
export default const MyComponent = (props) => {
  return `
    // do things
  `
}
</script>
```

### JSX Transformation
Custom walker for transforming JSX before WCC renders it down.
```js
const visitor = config.customWalker || defaultWalker || {};

walk.simple(acorn.parse(moduleContents, {
  ecmaVersion: 'latest',
  sourceType: 'module'
}), visitor, {
  ...walk.base
});
```

**Comment by thescientist13** [comment]
Would also be beneficial so as to keep transformation specific logic outside of core.  For example, even with JSX, we have to use `fs` to read off the filesystem, which won't get handled by custom loaders.
```js
function registerDependencies(moduleURL, definitions, depth = 0) {
  const moduleContents = fs.readFileSync(moduleURL, 'utf-8');
  const nextDepth = depth += 1;
  const customParser = getParser(moduleURL);
  const parser = customParser ? customParser.parser : acorn;
  const config = customParser ? customParser.config : {
    ...walk.base
  };

  walk.simple(acorn.parse(moduleContents, {
    ecmaVersion: 'latest',
    sourceType: 'module'
  }), {
     ...
  }, {
    ...walk.base
  });
}
```

So if this file ends in _.jsx_ then, it has to manually get turned into JSX.

TL;DR Custom loaders only help with transforming via ESM / `import`, how do we work around that?  Maybe everything should be `fetch` instead?  Maybe that would help with things like #36 ?
